README.md
# Shiny Affiliate

- Uses PHP class called AmazonProductApi to set proper parameters and create signed URL's for use with Amazon's Product Advertising API
- Also makes use of a very basic cURL class to retrieve response XML file
- Then, on one page, user can view products, add them to a remote cart (or remove something they've already added), and finally click the Purchase URL link to checkout at Amazon proper.

## Notes about Response Groups

When performing an ItemLookup operation to the Amazon Product Advertising API, you must specify Response Group(s) (hereafter 'RG'), which signify the type of information you want to get back.

Examples: 
  - Images RG will return URL's to images of a particular product, 
  - ItemAttributes RG will return product details like Manufacturer, Author, Title, etc.
  - Offers RG will (generally) return information on pricing and availability from particular merchants of a product

There are some exceptions, specifically when you want to display current pricing:
  - When multiple sizes or types of a product are for sale (Ex: clothing, shoes), the Offers RG will not return price information, you must instead use the Variations RG
  - There is not a way to display current pricing for Kindle eBooks through the Product Advertising API; you can show the List Price or other eBook information by using the ItemAttributes RG, but keep in mind that List Price is not the same as current selling price

Also note: **Digital items, like Kindle eBooks and mp3's, cannot be added to a remote shopping cart** [per Amazon](http://docs.aws.amazon.com/AWSECommerceService/latest/DG/ShoppingCartConcepts.html#ItemsThatCannotBeAddedtotheActiveCartArea)

## Additional Resources

[Product Advertising API Scratchpad](http://associates-amazon.s3.amazonaws.com/scratchpad/index.html) - great for testing which elements are returned by each RG

[Product Advertising API Quick Reference Card](http://s3.amazonaws.com/awsdocs/Associates/2011-08-01/prod-adv-api-qrc-2011-08-01.pdf) - lists required/optional parameters as well as relevant RG for each operation


CHANGELOG.md
## v.2.0.0 (November 25, 2013)
- Complete overhaul and redesign
  - Original page required postbacks after every addition/subtraction from the shopping cart, and performed Amazon cart requests on each postback as well
  - New logic uses Javascript and jQuery to add/remove from cart, and only calls 1 Amazon cart request to create the cart and forward the user to the checkout URL
  - Changed itemLookup operations to be configured based on an xml file in the filesystem (rather than hard-coded into the page), making updating products much easier
  - Redesigned page layout to take advantage of Bootstrap 3's new responsive design, and added visible cart showing contents (instead of just a link with a total amount)

## v.1.2.0 (June 8, 2013)
- Renamed AmazonApi to AmazonProductApi for clarity
- Added a changelog!
- Renamed method itemLookUpWithOffersImages to itemLookup which now takes comma-separated Response Groups as second parameter

## v.1.1.1 (April 3, 2013)
- Fixed bad logic when dealing with page refreshes (did not work correctly in all browsers)

## v.1.1.0 (March 19, 2013)
- Added methods to AmazonApi class:
  - itemLookUpWithOffersImages returns a signed URL for product information for a comma-separated list of [ASINs](http://en.wikipedia.org/wiki/Amazon_Standard_Identification_Number)
  - cartCreate returns a signed URL that builds a remote shopping cart with the specified items/quantities
  - cartAdd returns a signed URL that adds items to an existing remote shopping cart
  - cartModify returns a signed URL that allows items to be added or removed from an existing remote shopping cart
  - cartGet returns a signed URL that retrieves an existing remote shopping cart
- index.php posts-back to itself each time an add-to-cart or delete-from-cart button is clicked. Using session variables to keep track of the contents of the cart and cart identifiers, a decision tree is traversed to determine the correct operation to perform (create, add, modify, etc.). A cart checkout link appears when the user has a valid cart, which redirects user to Amazon to checkout.

## v.1.0.0
- **Initial release**
- Adapted AmazonApi class from the signed URL tutorial at: [webtutsdepot.com/2009/10/13/amazon-signed-request-php/](webtutsdepot.com/2009/10/13/amazon-signed-request-php/)

Curl.php
<?php
/**
 * Perform curl operations without having to remember all curl options
 * 
 * Method curlString found in the comments at http://www.php.net/manual/en/curl.examples.php
 * but this or similar code pieces I've seen at various places around the web. 
 */
class Curl
{
	/**
	 * Retrieves a file using curl and provides a string
	 * 
	 * @param string $url location of file
	 * @return string $output the file as a string from the url provided
	 */
	public function curlString($url)
	{
		// create curl resource
        $ch = curl_init();

        // set url
        curl_setopt($ch, CURLOPT_URL, $url);

        //return the transfer as a string
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);

        // $output contains the output string
        $output = curl_exec($ch);

        // close curl resource to free up system resources
        curl_close($ch);
		
		return $output;
	}
}
?>

index.php
<?php
session_start();
require_once 'AmazonProductApi.php';
require_once 'Curl.php';

$Amazon = new AmazonProductApi();
$Curl = new Curl();

// array of ASINs to get product info for
$items = array('0452284694','0072230665','0812550706','B00BUADSMQ','0765325950','B0096KEMUY');
// type of information we want to get back
$responseGroup = 'Offers,Images,ItemAttributes';
// get a signed URL to lookup product info; pull in info using curl
$queryUrl = $Amazon->itemLookup($items, $responseGroup);
$request = $Curl->curlString($queryUrl);
$response = simplexml_load_string($request);

// variable to hold signed URL from one of the operations below, variable for purchase link, and setting cart count
$opUrl = '';
$purchaseUrl = '';
$cartCount = isset($_SESSION['CartCount']) ? (int)$_SESSION['CartCount'] : 0;
// if POST['ASIN'] exists: either need to create a cart, add to an existing cart, or delete from a cart
if ( isset($_POST['asin']) )
{
	// For add, create and delete
	$searchTerm = substr($_POST['asin'], 1);
	$key = array_search($searchTerm, $items);
	// For add & create: set offerId array to correct item's id, and set qty array to 1 
	$offerId = array($key => $response->Items->Item[$key]->Offers->Offer->OfferListing->OfferListingId);
	$qty = array($key => 1);
	// if a cart ID exists (SESSION['CartId']), then add or delete from it
	if ( isset($_SESSION['CartId']) )
	{
		// if ASIN is in session variable (SESSION['a#']), delete from cart (cartModify()); Had to add 'a' to asin (can't use numeric-only key)
		$findAsin = 'a' . $items[$key];
		if ( isset($_SESSION[$findAsin]) )
		{
			// For delete: change qty to 0 and set cartItemId array to correct id from session variable
			$cartItemId = array($key => $_SESSION[$findAsin]);
			$qty[$key] = 0;
			$opUrl = $Amazon->cartModify($cartItemId, $qty, $_SESSION['HMAC'], $_SESSION['CartId']);
			// unset session variable for the ASIN and remove 1 from cart
			unset($_SESSION[$findAsin]);
			$cartCount--;
		}
		// else add to cart (cartAdd())
		else 
		{
			$opUrl = $Amazon->cartAdd($offerId, $qty, $_SESSION['HMAC'], $_SESSION['CartId']);
			$cartCount++;
		}
	}
	// if no cart ID exists, create a cart (createCart())
	else 
	{
		$opUrl = $Amazon->cartCreate($offerId, $qty);
		$cartCount++;
	}
}
// if this is not a POST, then it's first load; load default purchase buttons and unset opUrl and purchaseUrl
else 
{
	unset($opUrl, $purchaseUrl);
}

// if an operation is needed, load the response and set purchase url and session variables 
if ( isset($opUrl) )
{
	$cartRequest = $Curl->curlString($opUrl);
	$checkout = simplexml_load_string($cartRequest);
	$purchaseUrl = $checkout->Cart->PurchaseURL;
	$_SESSION['HMAC'] = (string)$checkout->Cart->HMAC;
	$_SESSION['CartId'] = (string)$checkout->Cart->CartId;
	$_SESSION['CartCount'] = $cartCount;
	// if cart has items, set ASIN session variables if item is in cart
	if ( $cartCount > 0 )
	{
		foreach ($checkout->Cart->CartItems->CartItem as $cartItem) 
		{
			$setAsin = 'a' . (string)$cartItem->ASIN;
			if ( !isset($_SESSION[$setAsin]) )
			{
				$_SESSION[$setAsin] = (string)$cartItem->CartItemId;
			}
		}
	}
	// otherwise last item removed from cart, delete cart purchase link
	else
	{
		unset($purchaseUrl);
	}
}
?>

<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>Store Test</title>
	<link href="//netdna.bootstrapcdn.com/twitter-bootstrap/2.3.2/css/bootstrap-combined.min.css" rel="stylesheet">
    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
  	<!-- NAVBAR -->
    <div class="navbar navbar-fixed-top navbar-inverse">  
      <div class="navbar-inner">  
        <div class="container">
          <a class="btn btn-navbar" data-toggle="collapse" data-target=".nav-collapse">
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
          </a>
          <ul class="nav">
            <li class="active"><a class="brand" href="../index.html">Test Store</a></li>
          </ul>
          <div class="nav-collapse collapse">
            <ul class="nav">   
              <li><a href="../index.html#about"><i class="icon-home icon-white"></i> About</a></li>  
              <li><a href="#"><i class="icon-book icon-white"></i> Store</a></li>
            <?php   // show cart checkout link if a purchase URL is available
            if ( isset($purchaseUrl) ): ?> 
              <li><a href="<?php echo $purchaseUrl; ?>" target="_blank"><i class="icon-shopping-cart icon-white"></i> Cart Checkout <span class="badge"><?php echo $cartCount ?></span></a></li>
            <?php endif; ?> 
            </ul>  
          </div>  
        </div>  
      </div>  
    </div>
    <!-- THUMBNAILS -->
    <div class="container store">
      <form method="post">
      	<h1 style="margin-top: 50px">Buy some great stuff from Amazon!</h1>
        <ul class="thumbnails">
<?php foreach ( $response->Items->Item as $item ):?>
			<li class="span4">
            	<div class="thumbnail">
					<img src="<?php echo $item->LargeImage->URL; ?>" width="200" alt="<?php echo $item->ItemAttributes->Title; ?> cover">
					<h5><?php echo $item->ItemAttributes->Title; ?></h5>
					<p>Amazon Price: 
						<?php echo $item->Offers->Offer->OfferListing->Price->FormattedPrice; ?>
					</p>
					<button type="submit" name="asin" value="<?php echo $currAsin = 'a' . $item->ASIN; ?>"
					<?php // set button value to ASIN#, and change button text based on existence of session variable
	                echo (isset($_SESSION[$currAsin]) ? 'class="btn btn-danger"><i class="icon-minus icon-white"></i> from ' : 
	                  'class="btn btn-success"><i class="icon-plus icon-white"></i> to '); ?>Amazon Cart</button>
				</div>
			</li>
<?php endforeach; ?>
		</ul>
	  </form>
	</div>
<script src="//netdna.bootstrapcdn.com/twitter-bootstrap/2.3.2/js/bootstrap.min.js"></script>
</body>

AmazonProductApi.php
<?php
/**
 * Use Amazon Product API to display products and product info from Amazon.com
 *
 * Uses Amazon's Product Advertising API to display products and product info
 * from Amazon using RESTful web requests; can also be used to build remote 
 * shopping carts to allow users to checkout directly from Amazon.com
 * @version 1.2.0
 */
class AmazonProductApi
{
	protected $_publicKey = 'xxxxxxxxxxxxxxxxxxxxxx';
	protected $_privateKey = 'xxxxxxxxxxxxxxxxxxxxxx';
	protected $_associateTag = 'xxxxxxxxxxxxxxxxxxxxxx';
	
	/**
	 * Get a signed URL
	 *  
	 * @param array $param used to build url
	 * @return array $signature returns the signed string and its components
	 */
	protected function generateSignature($param)
	{
		$signature['method'] = 'GET';
		$signature['host'] = 'webservices.amazon.com';
		$signature['uri'] = '/onca/xml';
		
		$param['Service'] = 'AWSECommerceService';
		$param['AWSAccessKeyId'] = $this->_publicKey;
		$param['AssociateTag'] = $this->_associateTag;
		$param['Timestamp'] = gmdate("Y-m-d\TH:i:s\Z");
		$param['Version'] = '2011-08-01';
		ksort($param);
		
		// URL encode key/value pairs
		foreach($param as $key => $value)
		{
			$key = str_replace("%7E", "~", rawurlencode($key));
			$value = str_replace("%7E", "~", rawurlencode($value));
			$queryParamsUrl[] = $key . '=' . $value;
		}
		
		// add ampersands between key/value pairs
		$signature['queryUrl'] = implode("&", $queryParamsUrl);
		
		// start signature creation
		$stringToSign = $signature['method'] . "\n" . $signature['host'] . "\n" . $signature['uri'] . "\n" . $signature['queryUrl'];
		$signature['string'] = str_replace("%7E", "~", rawurlencode(base64_encode(hash_hmac("sha256", $stringToSign, $this->_privateKey, TRUE))));
		return $signature;
	}
	
	/**
	 * Get signed url response
	 * 
	 * @param array $params
	 * @return string $signedUrl a query url with signature 
	 */
	 public function getSignedUrl($params)
	 {
	 	$signature = $this->generateSignature($params);
		return $signedUrl = "http://" . $signature['host'] . $signature['uri'] . '?' . $signature['queryUrl'] . '&Signature=' . $signature['string'];
	 }
	 
	 /**
	  * Get a signed url for ItemLookup Operation 
	  * 
	  * @param array $items a list of ASINs to lookup
	  * @param string $responseGroup a comma-separated string with desired response groups, defaults to Offers,Images
	  * @return string $signedUrl a query url with signature
	  */
	 public function itemLookup($items, $responseGroup = 'Offers,Images')
	 {
	 	$parameters = array(
		"ResponseGroup"=>$responseGroup,
		"Operation"=>"ItemLookup",
		"MerchantId"=>"Amazon",
		"IdType"=>"ASIN"
		);
	 	if ( is_array($items) )
		{
			$parameters['ItemId'] = implode(',', $items);
		}
	 	return $signedUrl = $this->getSignedUrl($parameters);
	 }
	 
	 /**
	  * Get a signed url for CartCreate Operation
	  * 
	  * @param array $offerId an array of OfferListingId's from ItemLookup xml file (or similar); keys should be ints
	  * @param array $qty the quantity of each offer to add to cart
	  * @return string $signedUrl a query url with signature
	  */
	 public function cartCreate($offerId, $qty)
	 {
	 	if ( is_array($offerId) && is_array($qty) )
		{
			$parameters = array(
			"Operation"=>"CartCreate"
			);
			foreach ( $offerId as $key => $value ) 
			{
				$parameters['Item.' . $key . '.OfferListingId'] = $value;
				$parameters['Item.' . $key . '.Quantity'] = $qty[$key];
			}
			return $signedUrl = $this->getSignedUrl($parameters);
		}
	 }
	 
	 /**
	  * Get signed url for CartAdd Operation
	  * 
	  * @param array $offerId an array of OfferListingId's from ItemLookup xml file (or similar); keys should be ints
	  * @param array $qty the quantity of each offer to add to cart
	  * @param string $hmac Hash Message Authentication Code returned by CartCreate that identifies a cart
	  * @param string $cartId Alphanumeric token returned by CartCreate that identifies a cart
	  * @return string $signedUrl a query url with signature
	  */
	 public function cartAdd($offerId, $qty, $hmac, $cartId)
	 {
	 	if ( is_array($offerId) && is_array($qty) )
		{
			$parameters = array(
			"Operation"=>"CartAdd",
			"HMAC"=>$hmac,
			"CartId"=>$cartId
			);
			foreach ( $offerId as $key => $value ) 
			{
				$parameters['Item.' . $key . '.OfferListingId'] = $value;
				$parameters['Item.' . $key . '.Quantity'] = $qty[$key];
			}
			return $signedUrl = $this->getSignedUrl($parameters);
		}
	 }
	 
	 /**
	  * Get signed url for CartModify Operation
	  * 
	  * @param array $cartItemId an array of CartItemId's from CartCreate response xml file; keys should be ints
	  * @param array $qty the quantity of each offer to change in cart (0 to delete)
	  * @param string $hmac Hash Message Authentication Code returned by CartCreate that identifies a cart
	  * @param string $cartId Alphanumeric token returned by CartCreate that identifies a cart
	  * @return string $signedUrl a query url with signature
	  */
	  public function cartModify($cartItemId, $qty, $hmac, $cartId)
	  {
	  	if ( is_array($cartItemId) && is_array($qty) )
		{
			$parameters = array(
			"Operation"=>"CartModify",
			"HMAC"=>$hmac,
			"CartId"=>$cartId
			);
			foreach ( $cartItemId as $key => $value ) 
			{
				$parameters['Item.' . $key . '.CartItemId'] = $value;
				$parameters['Item.' . $key . '.Quantity'] = $qty[$key];
			}
			return $signedUrl = $this->getSignedUrl($parameters);
		}
	  }
	  
	  /**
	   * Get signed url for CartGet Operation
	   * 
	   * @param string $hmac Hash Message Authentication Code returned by CartCreate that identifies a cart
	   * @param string $cartId Alphanumeric token returned by CartCreate that identifies a cart
	   * @return string $signedUrl a query url with signature
	   */
	  public function cartGet($hmac, $cartId)
	  {
		$parameters = array(
		"Operation"=>"CartGet",
		"HMAC"=>$hmac,
		"CartId"=>$cartId
		);

		return $signedUrl = $this->getSignedUrl($parameters);
	  }
}
?>

AmazonApi.php
<?php
/**
 * Use Amazon API to display products and product info 
 *
 * Found at webtutsdepot.com/2009/10/13/amazon-signed-request-php/ with some
 * alterations to variable/class names.  3-19-13 SB adding more methods
 * 
 */
class AmazonApi
{
	protected $_publicKey = 'xxxxxxxxxxxx';
	protected $_privateKey = 'xxxxxxxxxxxxxx';
	protected $_associateTag = 'fountless-20';
	
	/**
	 * Get a signed URL
	 *  
	 * @param array $param used to build url
	 * @return array $signature returns the signed string and its components
	 */
	protected function generateSignature($param)
	{
		$signature['method'] = 'GET';
		$signature['host'] = 'webservices.amazon.com';
		$signature['uri'] = '/onca/xml';
		
		$param['Service'] = 'AWSECommerceService';
		$param['AWSAccessKeyId'] = $this->_publicKey;
		$param['AssociateTag'] = $this->_associateTag;
		$param['Timestamp'] = gmdate("Y-m-d\TH:i:s\Z");
		$param['Version'] = '2011-08-01';
		ksort($param);
		
		foreach($param as $key => $value)
		{
			$key = str_replace("%7E", "~", rawurlencode($key));
			$value = str_replace("%7E", "~", rawurlencode($value));
			$queryParamsUrl[] = $key . '=' . $value;
		}
		
		// append all "params=value"s with ampersands
		$signature['queryUrl'] = implode("&", $queryParamsUrl);
		
		// start signature creation
		$stringToSign = $signature['method'] . "\n" . $signature['host'] . "\n" . $signature['uri'] . "\n" . $signature['queryUrl'];
		$signature['string'] = str_replace("%7E", "~", rawurlencode(base64_encode(hash_hmac("sha256", $stringToSign, $this->_privateKey, TRUE))));
		return $signature;
	}
	
	/**
	 * Get signed url response
	 * 
	 * @param array $params
	 * @return string $signedUrl a query url with signature 
	 */
	 public function getSignedUrl($params)
	 {
	 	$signature = $this->generateSignature($params);
		return $signedUrl = "http://" . $signature['host'] . $signature['uri'] . '?' . $signature['queryUrl'] . '&Signature=' . $signature['string'];
	 }
	 
	 /**
	  * Get a signed url for ItemLookup Operation with Offers Response Group
	  * 
	  * @param array $items a list of ASINs to lookup
	  * @return string $signedUrl a query url with signature
	  */
	 public function itemLookupWithOffersImages($items)
	 {
	 	$parameters = array(
		"ResponseGroup"=>"Offers,Images",
		"Operation"=>"ItemLookup",
		"IdType"=>"ASIN"
		);
	 	if ( is_array($items) )
		{
			$parameters['ItemId'] = implode(',', $items);
		}
	 	return $signedUrl = $this->getSignedUrl($parameters);
	 }
	 
	 /**
	  * Get a signed url for CartCreate Operation
	  * 
	  * @param array $offerId an array of OfferListingId's from ItemLookup xml file (or similar); keys should be ints
	  * @param array $qty the quantity of each offer to add to cart
	  * @return string $signedUrl a query url with signature
	  */
	 public function cartCreate($offerId, $qty)
	 {
	 	if ( is_array($offerId) && is_array($qty) )
		{
			$parameters = array(
			"Operation"=>"CartCreate"
			);
			foreach ( $offerId as $key => $value ) 
			{
				$parameters['Item.' . $key . '.OfferListingId'] = $value;
				$parameters['Item.' . $key . '.Quantity'] = $qty[$key];
			}
			return $signedUrl = $this->getSignedUrl($parameters);
		}
	 }
	 
	 /**
	  * Get signed url for CartAdd Operation
	  * 
	  * @param array $offerId an array of OfferListingId's from ItemLookup xml file (or similar); keys should be ints
	  * @param array $qty the quantity of each offer to add to cart
	  * @param string $hmac Hash Message Authentication Code returned by CartCreate that identifies a cart
	  * @param string $cartId Alphanumeric token returned by CartCreate that identifies a cart
	  * @return string $signedUrl a query url with signature
	  */
	 public function cartAdd($offerId, $qty, $hmac, $cartId)
	 {
	 	if ( is_array($offerId) && is_array($qty) )
		{
			$parameters = array(
			"Operation"=>"CartAdd",
			"HMAC"=>$hmac,
			"CartId"=>$cartId
			);
			foreach ( $offerId as $key => $value ) 
			{
				$parameters['Item.' . $key . '.OfferListingId'] = $value;
				$parameters['Item.' . $key . '.Quantity'] = $qty[$key];
			}
			return $signedUrl = $this->getSignedUrl($parameters);
		}
	 }
	 
	 /**
	  * Get signed url for CartModify Operation
	  * 
	  * @param array $cartItemId an array of CartItemId's from CartCreate response xml file; keys should be ints
	  * @param array $qty the quantity of each offer to change in cart (0 to delete)
	  * @param string $hmac Hash Message Authentication Code returned by CartCreate that identifies a cart
	  * @param string $cartId Alphanumeric token returned by CartCreate that identifies a cart
	  * @return string $signedUrl a query url with signature
	  */
	  public function cartModify($cartItemId, $qty, $hmac, $cartId)
	  {
	  	if ( is_array($cartItemId) && is_array($qty) )
		{
			$parameters = array(
			"Operation"=>"CartModify",
			"HMAC"=>$hmac,
			"CartId"=>$cartId
			);
			foreach ( $cartItemId as $key => $value ) 
			{
				$parameters['Item.' . $key . '.CartItemId'] = $value;
				$parameters['Item.' . $key . '.Quantity'] = $qty[$key];
			}
			return $signedUrl = $this->getSignedUrl($parameters);
		}
	  }
	  
	  /**
	   * Get signed url for CartGet Operation
	   * 
	   * @param string $hmac Hash Message Authentication Code returned by CartCreate that identifies a cart
	   * @param string $cartId Alphanumeric token returned by CartCreate that identifies a cart
	   * @return string $signedUrl a query url with signature
	   */
	  public function cartGet($hmac, $cartId)
	  {
		$parameters = array(
		"Operation"=>"CartGet",
		"HMAC"=>$hmac,
		"CartId"=>$cartId
		);

		return $signedUrl = $this->getSignedUrl($parameters);
	  }
}
?>
